pkg/web/apps/apps_psql_repository.go

```diff
@@ -19,16 +22,59 @@ func NewPostgreSQLRepository(db *sql.DB) Repository {
 
 // GetApps retrieves all apps from the database
 func (a *appsPostgreSQLRepository) GetApps() (*[]models.App, error) {
-	app1 := models.App{
-		ID:                    "a0874c82-2b7f-11e9-b210-d663bd873d93",
-		AppID:                 "com.aerogear.app1",
-		AppName:               "app1",
-		NumOfDeployedVersions: 1,
-		NumOfClients:          1,
-		NumOfAppLaunches:      1,
+	rows, err := a.db.Query(`SELECT id,app_id,app_name FROM app WHERE deleted_at IS NULL;`)
+
+	if err != nil {
+		log.Error(err)
+		return nil, models.ErrInternalServerError
 	}
 
-	apps := []models.App{app1}
+	defer func() {
+		err := rows.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	apps := []models.App{}
+	for rows.Next() {
+		var a models.App
+		if err = rows.Scan(&a.ID, &a.AppID, &a.AppName); err != nil {
+			log.Error(err)
+		}
+
+		apps = append(apps, a)
+	}
 
 	return &apps, nil
 }
+
+func (a *appsPostgreSQLRepository) GetAppVersions(id string) (*[]models.Version, error) {
+	rows, err := a.db.Query(`SELECT id,version,app_id,disabled,disabled_message,num_of_clients,num_of_app_launches FROM version WHERE app_id = $1`, id)
+
+	if err != nil {
+		log.Error(err)
+		return nil, models.ErrInternalServerError
+	}
+
+	defer func() {
+		err := rows.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	versions := []models.Version{}
+
+	// iterate over the rows and add the data to the array of versions
```

Comment:
Hi @craicoverflow,

You don't need to do all these interactions. You can fetch all the data in just one query doing something as follows.  

```sql
  SELECT
     app.id as ID,
     app.app_id as AppID,
     app.app_name as AppName,
     (select count(DISTINCT v1.version_id) from version v1 where v1.app_id = app.app_id) as NumOfDeployedVersions,
     (select count(DISTINCT d1.device_id) from device d1 where d1.version_id in (
        SELECT DISTINCT v2.version_id FROM version v2 WHERE v2.app_id = app.app_id)
     ) as NumOfClients,
     (select sum(v3.num_app_launches) from version v3 where v3.app_id = app.app_id) as NumOfAppLaunches
    FROM
      app 
    WHERE
      app.deleted_at IS NOT NULL 
```
Also, you can do it with a better performance by using INNER JOINS. See [here](https://www.postgresql.org/docs/8.3/tutorial-join.html)

Reply:
Hi @camilamacedo86 you are right. I have updated this now. One thing though - INNER JOINS do not work for this, if a row in the app table does not have any associated versions then the query does not return that app at all. For this reason we must use LEFT join.

Reply:
Hi @craicoverflow,

Yes. You have LEFT JOIN and OUT JOIN and INNER JOIN. Regards, this scenario I believe that you are correct and it will need the LEFT JOIN. 

Note that my comment was to guide you to look for how you can solve it in the database instead of doing this interaction then since you reply it hows that the goal was achieved. :-)